ecosystem/purescript-0.14-migration-guide.md

```diff
@@ -0,0 +1,622 @@
+
+```purs
+Builder.build (Builder.flip Builder.merge ...)
+```
+
+#### The `Data.Either.fromLeft` and `Data.Either.fromRight` functions are now total functions.
+
+Implemented in [purescript-either#48](https://github.com/purescript/purescript-either/pull/48) with additional discussion in [purescript-either#47](https://github.com/purescript/purescript-either/issues/47) and [purescript-either#56](https://github.com/purescript/purescript-either/issues/56).
+
+The `Data.Either.fromLeft` and `Data.Either.fromRight` functions are now total functions and take a default value (like `Data.Maybe.fromMaybe`) instead of partial functions (like `Data.Maybe.fromJust`).
+
+**To fix:**
+
+The most common way to use `fromLeft` and `fromRight` is to unsafely discharge the `Partial` constraint using `unsafePartial` like this:
+
+```purs
+x = unsafePartial fromLeft $ Left 10
+y = unsafePartial fromRight $ Right "hi"
+```
+
+This can be replaced with the new lazy `fromLeft'` and `fromRight'` combined with `unsafeCrashWith`:
+
+```purs
+x = fromLeft' (\_ -> unsafeCrashWith "Unexpected Right") $ Left 10
+x = fromRight' (\_ -> unsafeCrashWith "Unexpected Left") $ Right "hi"
+```
+
+Alternately, you can provide a default value in order to use `fromLeft` and `fromRight` safely.
+
+#### The `quickcheck` library now uses `NonEmptyArray` and `NonEmptyList` instead of `NonEmpty Array` and `NonEmpty List`.
+
+Implemented in [purescript-quickcheck#118](https://github.com/purescript/purescript-quickcheck/pull/118) with additional discussion in [purescript-quickcheck#109](https://github.com/purescript/purescript-quickcheck/issues/109).
+
+The `NonEmpty Array` type predates the `NonEmptyArray` type, which provides O(n) conversion to and from `Array`. The QuickCheck library has switched to use `NonEmptyArray` and `NonEmptyList`, which affects libraries which create generators using functions like `elements` or `frequency`.
+
+**To fix:**
+
+Instead of constructing a `NonEmpty Array` for functions like `Test.QuickCheck.Gen.elements`, you should instead construct a `NonEmptyArray`.
+
+Here's a common way to construct a `NonEmpty Array` in an `Arbitrary` instance:
+
+```purs
+import Data.NonEmpty ((:|))
+
+instance arbitraryOrdering :: Arbitrary Ordering where
+  arbitrary = elements $ LT :| [ EQ, GT ]
+```
+
+This can be replaced by this instance, to minimize the difference between the old code and the new code:
+
+```purs
+import Data.Array.NonEmpty (cons')
+
+instance arbitraryOrdering :: Arbitrary Ordering where
+  arbitrary = elements $ cons' LT [ EQ, GT ]
+```
+
+Or, for a slightly more performant version which constructs the array and converts it without an intermediate `cons`:
+
+```purs
+import Data.Array.NonEmpty (fromArray)
+
+instance arbitraryOrdering :: Arbitrary Ordering where
+  arbitrary = unsafePartial fromJust $ fromArray [ LT, EQ, GT ]
```

Comment:
Don’t we miss a call to `elements` here?

```suggestion
  arbitrary = elements $ unsafePartial fromJust $ fromArray [ LT, EQ, GT ]
```

Reply:
Yes! Good catch.